### ISSUE TITLE
track: `Files API + MFS`

### ISSUE THREAD

**Issue by diasdavid** [ipfs/js-ipfs#60]
The `files` API offers an interface to work with unixfs, a Merklelized Data Structure that represents a typical FileSystem used in a Unix OS. with files, dirs and symlinks. This API is what enables IPFS to add and retrieve files without having to ask the users to convert their data into a graph data structure.
#### To learn more about the Files API, you can
- watch https://www.youtube.com/watch?v=FX_AXNDsZ9k
- check files (mfs) implementation in go https://github.com/ipfs/go-ipfs/tree/master/mfs
- check unixfs implementation in go https://github.com/ipfs/go-ipfs/tree/master/unixfs
#### Implementation roadmap
- [x] Files API spec https://github.com/ipfs/specs/issues/88
- [x] Unixfs Spec 
- [x] jsipfs files add (current ipfs add)
  - https://github.com/ipfs/js-ipfs/pull/67
- [x] jsipfs files cat (current ipfs cat)
  - [x] cat subfiles 
- [x] jsipfs files get (current ipfs get)
- [ ] jsipfs files mv <source> <dest>  - Move files.
- [ ] jsipfs files cp <source> <dest>  - Copy files into mfs.
- [ ] jsipfs files ls <path>           - List directories.
- [ ] jsipfs files mkdir <path>        - Make directories.
- [ ] jsipfs files stat <path>         - Display file status.
- [ ] jsipfs files rm <path>...        - Remove a file.
- [ ] jsipfs files read <path>         - Read a file in a given mfs. 
- [ ] jsipfs files write <path> <data> - Write to a mutable file in a given filesystem.

**Comment by diasdavid** [comment]
**Updated this issue**

**Comment by nginnever** [comment]
<h3>jsipfs files get</h3>

Taking a note from @noffle and a stab at ```jsipfs files get```

Relevant go-ipfs links: 
https://github.com/ipfs/go-ipfs/blob/master/core/commands/get.go

<h3>ARGUMENTS:</h3>

    <ipfs-path> - The path to the IPFS object(s) to be outputted

<h3>OPTIONS:</h3>

    -o, --output            string - The path where output should be stored
    -a, --archive           bool   - Output a TAR archive
    -C, --compress          bool   - Compress the output with GZIP compression
    -l, --compression-level int    - The level of compression (1-9)

<h3>STEPS</h3>
1. Get the gzip compression options, bail if err (TODO: Look into js gzip modules)
   i. return NoCompression if compress bool is false
   ii. return DefaultCompression if compress bool is true but compress level is undefined
   iii. return NoCompression and error if bool is true and compress level is not between 1-9 
   iv. return compress level if none of the above conditions are met
2.  <a href="https://github.com/ipfs/go-ipfs/blob/master/commands/request.go#L48-L57">Get</a> the <a href="https://github.com/ipfs/go-ipfs/blob/master/core/core.go#L81-L118">node</a> from the command request context
3. Get the path from the argument 
4. <a href="https://github.com/ipfs/go-ipfs/blob/master/core/pathresolver.go#L23-L58">Resolve</a> the path.

**Comment by nginnever** [comment]
<h3>jsipfs files cat</h3>

A look at ```jsipfs files cat```

Relevant go-ipfs links: 
https://github.com/ipfs/go-ipfs/blob/master/core/commands/cat.go
https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/cat.go

<h3>ARGUMENTS:</h3>

    <ipfs-path>... - The path to the IPFS object(s) to be outputted

<h3>DESCRIPTION:</h3>

    Retrieves the object named by <ipfs-or-ipns-path> and outputs the data
    it contains.

<h3>STEPS</h3>

1. Get the node from the command request context.
2. If node.onlineMode = false, <a href="https://github.com/ipfs/go-ipfs/blob/master/core/core.go#L512-L528">setupOfflineRouting()</a>
     - TODO: List setupOfflineRouting steps
3. call cat() function passing in the context, the node, and the path from the arguments
     - create read streams array for each path provided
     - for each path, call <a href="https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/cat.go#L10-L16">coreunix.Cat(ctx, node, fpath)</a>
          - <a href="https://github.com/ipfs/go-ipfs/blob/master/core/pathresolver.go#L23-L58">resolve</a> the dagnode
          - return a readable stream 
     - append the stream from resolve to the  read streams array
     - collect the total length by adding each buffer length returned from resolve
     - return the readers array, length, and null or error
4. Do a <a href="https://github.com/ipfs/go-ipfs/blob/master/core/corerepo/gc.go#L153-L159">conditional GC</a>
5.  set the length in the response
6. combine the readers streams into one read stream
7. check if the length < the smallest progress bar size
8. create the progress bar if not
9. set the res to the read stream.

**Comment by diasdavid** [comment]
merged https://github.com/ipfs/js-ipfs/pull/67

**Comment by diasdavid** [comment]
@nginnever I know you are working on this. Can you make a PR tagged with WIP from a branch under this project?

**Comment by diasdavid** [comment]
-> https://github.com/ipfs/js-ipfs/pull/156

**Comment by diasdavid** [comment]
The remaining items (specific to mfs) go into the Milestone 6 - Feature parity

**Comment by diasdavid** [comment]
//cc @VictorBjelkholm

**Comment by diasdavid** [comment]
Now Files Add work as expected with the option to do trickle-dag too! https://github.com/ipfs/js-ipfs/pull/707

**Comment by diasdavid** [comment]
cat/get subtrees is now there too :)

**Comment by diasdavid** [comment]
Now tracked here https://github.com/ipfs/ipfs/blob/master/IMPLEMENTATION_STATUS.md

tl;dr; Files is done, MFS isn't
